src/main/java/ca/on/oicr/pineryreports/reports/impl/LanesBillingReport.java

```diff
@@ -0,0 +1,464 @@
+            : RNA_LANE;
+        
+        for (Map.Entry<String, Integer> entry : projectsInLane.entrySet()) {
+          BigDecimal projectPercent = getPercentProjectInLane(entry.getValue(), samplesInLane);
+          DetailedObject newReport = new DetailedObject(completionDate, instrumentName, instrumentModel, runName, 
+              Integer.toString(laneNumber), entry.getKey(), projectPercent, runStatus, laneContents);
+          detailedData.add(newReport);
+          addSummaryData(summaryData, newReport);
+        }
+      }
+    }
+    return new ReportObject(detailedData, summaryData);
+  }
+  
+  private void addSummaryData(Map<String, Map<String, BigDecimal>> summary, DetailedObject newReport) {
+    String summaryKey = getSummaryKey(newReport);
+    if (summary.containsKey(summaryKey)) {
+      // increment lane number+type for this project+instrument
+      addSummaryLanes(summary.get(summaryKey), newReport.getLaneContents(), newReport.getProjLibsPercent());
+    } else {
+      // add new project+instrument and lane number+type
+      Map<String, BigDecimal> newLanes = makeNewSummaryLanes(newReport.getLaneContents(), newReport.getProjLibsPercent());
+      summary.put(summaryKey, newLanes);
+    }
+  }
+  
+  private BigDecimal getPercentProjectInLane(Integer numProjectLibraries, Integer numLaneLibraries) {
+    return new BigDecimal(numProjectLibraries).divide(new BigDecimal(numLaneLibraries), 1,  RoundingMode.HALF_UP);
+  }
+  
+  private String getInstrumentModel(Integer instrumentId, Map<Integer, InstrumentDto> instruments, Map<Integer, InstrumentModelDto> models) {
+    return models.get(
+        instruments.get(instrumentId).getModelId())
+      .getName();
+  }
+  
+  private String getInstrumentName(Integer instrumentId, Map<Integer, InstrumentDto> instrumentsById) {
+    InstrumentDto instrument = instrumentsById.get(instrumentId);
+    return instrument == null ? "Unknown" : instrument.getName();
+  }
+  
+  private String getSummaryKey(DetailedObject row) {
+    return row.getProject() + ":" + row.getInstrumentModel();
+  }
+  
+  private Map<String, BigDecimal> makeNewSummaryLanes(String laneType, BigDecimal laneCount) {
+    Map<String, BigDecimal> lanes = new HashMap<String, BigDecimal>();
+    lanes.put(DNA_LANE, BigDecimal.ZERO);
+    lanes.put(RNA_LANE, BigDecimal.ZERO);
+    lanes.put(MIXED_LANE, BigDecimal.ZERO);
+    addSummaryLanes(lanes, laneType, laneCount);
+    return lanes;
+  }
+  
+  private void addSummaryLanes(Map<String, BigDecimal> lanes, String laneType, BigDecimal laneCount) {
+    lanes.merge(laneType, laneCount, BigDecimal::add);
+  }
+  
+  static final List<Map<String, Map<String, BigDecimal>>> listifySummary(Map<String, Map<String, BigDecimal>> summary) {
+    // need to convert it to a list, because getRow() takes an index and the treemap doesn't yet have one of those
+    List<Map<String, Map<String, BigDecimal>>> regrettable = new ArrayList<Map<String, Map<String, BigDecimal>>>();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace the type specification in this constructor call with the diamond operator ("<>"). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)